src/extract.rs

```diff
@@ -0,0 +1,76 @@
+use std::ops::RangeBounds;
+
+use crate::TableOption;
+
+/// Returns a new [Table] that reflects a segment of the referenced [Table]
+///
+/// The segment is defined by [RangeBounds<usize>] for Rows and Columns
+///
+/// ```rust,no_run
+/// Extract::new(rows: .., columns: ..)
+/// ```
+/// 
+/// # Example
+///
+/// ```
+/// use tabled::{Table, Format, Row, Modify};
+///
+/// let data = vec![
+///     (0, "Grodno", true),
+///     (1, "Minsk", true),
+///     (2, "Hamburg", false),
+///     (3, "Brest", true),
+/// ];
+///
+/// let table = Table::new(&data)
+///                .with(Modify::new(Row(1..)).with(Format(|s| format!(": {} :", s))))
+///                .with(Extract::new(1..=2, 1..))
+///                .to_string();
+///
+/// assert_eq!(table, "+-------------+-----------+\n\
+///                    | : Grodno :  | : true :  |\n\
+///                    +-------------+-----------+\n\
+///                    |  : Minsk :  | : true :  |\n\
+///                    +-------------+-----------+\n");
+///
+/// assert_ne!(table, "+-------+-------------+-----------+\n\
+///                    |  i32  |    &str     |   bool    |\n\
+///                    +-------+-------------+-----------+\n\
+///                    | : 0 : | : Grodno :  | : true :  |\n\
+///                    +-------+-------------+-----------+\n\
+///                    | : 1 : |  : Minsk :  | : true :  |\n\
+///                    +-------+-------------+-----------+\n\
+///                    | : 2 : | : Hamburg : | : false : |\n\
+///                    +-------+-------------+-----------+\n\
+///                    | : 3 : |  : Brest :  | : true :  |\n\
+///                    +-------+-------------+-----------+\n");
+/// ```
+///
+pub struct Extract<R, C>
+where
+    R: RangeBounds<usize>,
+    C: RangeBounds<usize>,
+{
+    rows: R,
+    columns: C,
+}
+
+impl<R, C> Extract<R, C>
+where
+    R: RangeBounds<usize>,
+    C: RangeBounds<usize>,
+{
+    pub fn new(rows: R, columns: C) -> Extract<R, C> {
+        Extract { rows, columns }
+    }
```

Comment:
Maybe our current interface with `new` function is a bit ugly?

How do you feel about the next interface.

```rust
    // rows from 0 to max
    pub fn columns(columns: C) -> Extract<R, C>;
    // columns from 0 to max
    pub fn rows(rows: R) -> Extract<R, C>;
    pub fn segment(rows: R, columns: C) -> Extract<R, C>;
```

Reply:
I will experiment shifting the implementation to be functional as you suggest. 

This will most likely correct the issue mentioned below with access to the `papergrid` extract api being locked behind a reference to `self` within the TableOption change function.

Reply:
The change has been made. Tests have been added/rewritten to accommodate this change.

Took a stab at adding doc references using this resource: [rustdocs](https://doc.rust-lang.org/rustdoc/linking-to-items-by-name.html#valid-links)

new apis:
- Extract::columns
- Extract::rows
- Extract::segment